.github/workflows/upgrade-testing-run.yml

```diff
@@ -0,0 +1,94 @@
+##
+# A callable workflow that runs WordPress upgrade testing under the conditions provided.
+##
+name: Upgrade Tests
+
+on:
+  workflow_call:
+    inputs:
+      os:
+        description: 'Operating system to run tests on'
+        required: false
+        type: 'string'
+        default: 'ubuntu-latest'
+      wp:
+        description: 'The version of WordPress to start with.'
+        required: true
+        type: 'string'
+      new-version:
+        description: 'The version of WordPress to update to.'
+        type: 'string'
+      php:
+        description: 'The version of PHP to use, in the format of X.Y'
+        required: true
+        type: 'string'
+      multisite:
+        description: 'Whether to run tests as multisite'
+        required: false
+        type: 'boolean'
+        default: false
+      db-type:
+        description: 'Database type. Valid types are mysql and mariadb'
+        required: false
+        type: 'string'
+        default: 'mysql'
+      db-version:
+        description: 'Database version'
+        required: false
+        type: 'string'
+        default: '5.7'
+
+jobs:
+  # Runs upgrade tests on a build of WordPress.
+  #
+  # Performs the following steps:
+  # - Checks out the repository.
+  # - Sets up PHP.
+  # - Starts the database server.
+  # - Downloads the specified version of WordPress.
+  # - Creates a `wp-config.php` file.
+  # - Installs WordPress.
+  # - Updates to the latest minor version.
+  # - Updates to the version of WordPres being tested.
+  upgrade-tests:
+    name: ${{ inputs.wp }} to ${{ inputs.new-version && inputs.new-version || 'latest' }} / PHP ${{ inputs.php }} with ${{ 'mariadb' == inputs.db-type && 'MariaDB' || 'MySQL' }} ${{ inputs.db-version }}${{ inputs.multisite && ' multisite' || '' }}
+    runs-on: ${{ inputs.os }}
+    timeout-minutes: 20
+
+    services:
+      database:
+        image: ${{ inputs.db-type }}:${{ inputs.db-version }}
+        ports:
+          - 3306
+        options: --health-cmd="mysqladmin ping" --health-interval=30s --health-timeout=10s --health-retries=5 -e MYSQL_ROOT_PASSWORD=root -e MYSQL_DATABASE=test_db --entrypoint sh ${{ inputs.db-type }}:${{ inputs.db-version }} -c "exec docker-entrypoint.sh mysqld --default-authentication-plugin=mysql_native_password"
+
+    steps:
+      - name: Checkout the repository
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c
+
+      - name: Set up PHP ${{ inputs.php }}
+        uses: shivammathur/setup-php@d30ad8b1843ace22e6698ab99bbafaa747b6bd0d # v2.24.0
+        with:
+          php-version: '${{ inputs.php }}'
+          coverage: none
+          tools: wp-cli
+
+      - name: Start the database server
+        run: |
+          sudo systemctl start ${{ inputs.db-type }}
+
+      - name: Download WordPress ${{ inputs.wp }}
```

Comment:
I separated all of these commands into their own steps. I try to make each step as short as possible. It makes it easier to see where failures occur within the workflow.